File: framework/languages/dotnet/e2e/RegistryMessaging/Server/Provider/Configuration/ServerOptions.cs
using RegistryMessaging.Server.Provider.Endpoints;
using RegistryMessaging.Server.Provider.Handlers;
using RegistryMessaging.Server.Provider.Infrastructure;
using RegistryMessaging.Server.Provider;
namespace RegistryMessaging.Server.Provider.Configuration;

internal sealed record ServerOptions(
    string Role,
    string HttpUrl,
    string LogDir,
    string? EvidenceFile,
    string Rid,
    string? RegistryPubEndpoint,
    string? RegistryRouterEndpoint,
    string? ChannelEndpoint,
    string? ManualClientEndpoint,
    string? WorkflowEndpoint,
    string? RouteEndpoint,
    int Weight,
    IReadOnlyList<string> RoutePeers)
{
    public static ServerOptions Parse(string[] args, string defaultRole)
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var routePeers = new List<string>();

        for (var i = 0; i < args.Length; i++)
        {
            var key = args[i];
            if (!key.StartsWith("--", StringComparison.Ordinal))
            {
                continue;
            }

            if (i + 1 >= args.Length)
            {
                throw new ArgumentException($"Missing value for {key}.");
            }

            var value = args[++i];
            if (key == "--route-peer")
            {
                routePeers.Add(value);
            }
            else
            {
                values[key[2..]] = value;
            }
        }

        var rid = values.GetValueOrDefault("rid", "node");
        Environment.SetEnvironmentVariable("ZLINK_E2E_RID", rid);
        return new ServerOptions(
            defaultRole,
            values.GetValueOrDefault("http-url", "http://127.0.0.1:0"),
            values.GetValueOrDefault("log-dir", Path.Combine(Path.GetTempPath(), "zlink-dotnet-e2e-log")),
            values.GetValueOrDefault("evidence-file"),
            rid,
            values.GetValueOrDefault("registry-pub-endpoint"),
            values.GetValueOrDefault("registry-router-endpoint"),
            values.GetValueOrDefault("channel-endpoint"),
            values.GetValueOrDefault("manual-client-endpoint"),
            values.GetValueOrDefault("workflow-endpoint"),
            values.GetValueOrDefault("route-endpoint"),
            int.TryParse(values.GetValueOrDefault("weight"), out var weight) ? weight : 100,
            routePeers);
    }
}

File: framework/languages/node/e2e/RegistryMessaging/Server/Provider/Configuration/server-options.ts
export interface ServerOptions {
  readonly role: string;
  readonly httpUrl: string;
  readonly logDir: string;
  readonly evidenceFile?: string;
  readonly rid: string;
  readonly redisEndpoint?: string;
  readonly redisKeyPrefix?: string;
  readonly channelEndpoint?: string;
  readonly manualClientEndpoint?: string;
  readonly routeEndpoint?: string;
  readonly routePeers: readonly string[];
  readonly weight: number;
  readonly maxMessageSize: number;
}

export function parseServerOptions(args: readonly string[], defaultRole = 'provider'): ServerOptions {
  const values = new Map<string, string>();
  const routePeers: string[] = [];
  for (let i = 0; i < args.length; i += 1) {
    const key = args[i];
    if (!key.startsWith('--')) {
      continue;
    }
    if (i + 1 >= args.length) {
      throw new Error(`Missing value for ${key}.`);
    }
    const value = args[++i];
    if (key === '--route-peer') {
      routePeers.push(value);
    } else {
      values.set(key.slice(2), value);
    }
  }

  const rid = values.get('rid') ?? 'node';
  process.env.ZLINK_E2E_RID = rid;
  return {
    role: defaultRole,
    httpUrl: values.get('http-url') ?? 'http://127.0.0.1:0',
    logDir: values.get('log-dir') ?? '/tmp/zlink-node-e2e-log',
    evidenceFile: values.get('evidence-file'),
    rid,
    redisEndpoint: values.get('redis-endpoint'),
    redisKeyPrefix: values.get('redis-key-prefix'),
    channelEndpoint: values.get('channel-endpoint'),
    manualClientEndpoint: values.get('manual-client-endpoint'),
    routeEndpoint: values.get('route-endpoint'),
    routePeers,
    weight: parseWeight(values.get('weight')),
    maxMessageSize: parseMaxMessageSize(values.get('max-message-size'))
  };
}

function parseWeight(value: string | undefined): number {
  if (value === undefined) {
    return 100;
  }
  const weight = Number(value);
  if (!Number.isInteger(weight) || weight < 0 || weight > 100) {
    throw new Error('--weight must be an integer between 0 and 100.');
  }
  return weight;
}

function parseMaxMessageSize(value: string | undefined): number {
  if (value === undefined) {
    return 0;
  }
  const size = Number(value);
  if (!Number.isInteger(size) || size < 0) {
    throw new Error('--max-message-size must be a non-negative integer.');
  }
  return size;
}
